core/src/main/java/com/graphhopper/storage/index/LocationIndexTree.java

```diff
@@ -564,6 +572,133 @@ protected boolean check(int node, double normedDist, int wayIndex, EdgeIteratorS
+        }
+        
+        final double returnAllResultsWithin = distCalc.calcNormalizedDist(radius);
+        // implement a cheap priority queue via List, sublist and Collections.sort
+        final List<QueryResult> queryResults = new ArrayList<QueryResult>();
+        GHIntHashSet set = new GHIntHashSet();
+
+        for (int iteration = 0; iteration < _maxIterations; iteration++) {
+            // TODO: should we use the return value of earlyFinish?
+            findNetworkEntries(queryLat, queryLon, set, iteration);
+            final GHBitSet exploredNodes = new GHTBitSet(new GHIntHashSet(set));
+            final EdgeExplorer explorer = graph.createEdgeExplorer(edgeFilter);
+            set.forEach(new IntPredicate() {
+                @Override
+                public boolean apply(int node) {
+                    new XFirstSearchCheck(queryLat, queryLon, exploredNodes, edgeFilter) {
+                        @Override
+                        protected double getQueryDistance() {
+                            // do not skip search if distance is 0 or near zero (equalNormedDelta)
+                            return Double.MAX_VALUE;
+                        }
+
+                        @Override
+                        protected boolean check(int node, double normedDist, int wayIndex,
+                                EdgeIteratorState edge, QueryResult.Position pos) {
+                            if (normedDist < returnAllResultsWithin || queryResults.isEmpty()
+                                    || queryResults.get(0).getQueryDistance() > normedDist) {
+                                // TODO: refactor below:
+                                // - should only add edges within search radius (below allows the
+                                // returning of a candidate outside search radius if it's the only
+                                // one). Removing this test would simplify it a lot and probably
+                                // match the expected behaviour (see method description)
+                                // - create QueryResult first and the add/set as required - clean up
+                                // the index tracking business.
+                                int index = -1;
+                                for (int qrIndex = 0; qrIndex < queryResults.size(); qrIndex++) {
+                                    QueryResult qr = queryResults.get(qrIndex);
+                                    // overwrite older queryResults which are potentially more far
+                                    // away than returnAllResultsWithin
+                                    if (qr.getQueryDistance() > returnAllResultsWithin) {
+                                        index = qrIndex;
+                                        break;
+                                    }
+
+                                    // avoid duplicate edges
+                                    if (qr.getClosestEdge().getEdge() == edge.getEdge()) {
```

Comment:
Why aren't we using a set/map here and filter potential unwanted results after the for loop?

With for loop I mean:
`for (int iteration = 0; iteration < _maxIterations; iteration++) {`